File: editor/src/main/java/com/raredev/vcspace/editor/CommentSystem.java
package com.raredev.vcspace.editor;

import com.raredev.vcspace.utils.ToastUtils;
import io.github.rosemoe.sora.text.Content;
import io.github.rosemoe.sora.text.Cursor;
import org.eclipse.tm4e.languageconfiguration.model.CharacterPair;
import org.eclipse.tm4e.languageconfiguration.model.CommentRule;

public class CommentSystem {

  public static void addSingleComment(CommentRule commentRule, Content text, int line) {
    String commentLine = commentRule.lineComment;

    if (commentLine != null) {
      // Insert single-line comment at the beginning of the line
      var lineText = text.getLine(line).toString();
      var column = lineText.indexOf(lineText.trim().charAt(0))-1;
      
      text.insert(line, column, commentLine);
    }
  }

  public static void addBlockComment(CommentRule commentRule, Content text, Cursor cursor) {
    if (!cursor.isSelected()) {
      ToastUtils.showShort("No text selected", ToastUtils.TYPE_ERROR);
      return;
    }

    CharacterPair blockComment = commentRule.blockComment;
    if (blockComment != null) {
      String openPrefix = blockComment.open;
      String closePrefix = blockComment.close;

      if (openPrefix == null || closePrefix == null) {
        for (int line = cursor.getLeftLine(); line <= cursor.getRightLine(); line++) {
          addSingleComment(commentRule, text, line);
        }
        return;
      }

      // Insert multi-line comment at the beginning of the start line
      text.insert(cursor.getLeftLine(), cursor.getLeftColumn(), openPrefix);

      // Insert multi-line comment end at the end of the end line
      text.insert(cursor.getRightLine(), cursor.getRightColumn(), closePrefix);
    }
  }

  /*
  public static String uncomment(Content text) {
    StringBuilder sb = new StringBuilder();
    boolean inBlockComment = false;
    int i = 0;

    while (i < text.length()) {
      char c = text.charAt(i);

      if (inBlockComment) {
        if (c == '*' && i < text.length() - 1 && text.charAt(i + 1) == '/') {
          inBlockComment = false;
          i++;
        }
      } else {
        if (c == '/' && i < text.length() - 1 && text.charAt(i + 1) == '/') {
          i = text.toString().indexOf('\n', i + 2);
          if (i == -1) {
            break;
          }
        } else if (c == '/' && i < text.length() - 1 && text.charAt(i + 1) == '*') {
          inBlockComment = true;
          i++;
        } else {
          sb.append(c);
        }
      }

      i++;
    }

    return sb.toString();
  }
  */
}

File: editor/src/main/java/com/raredev/vcspace/editor/CommentSystem.kt
/*
 * This file is part of Visual Code Space.
 *
 * Visual Code Space is free software: you can redistribute it and/or modify it under the terms of
 * the GNU General Public License as published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Visual Code Space is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with Visual Code Space.
 * If not, see <https://www.gnu.org/licenses/>.
 */

package com.raredev.vcspace.editor

import io.github.rosemoe.sora.text.Content
import io.github.rosemoe.sora.text.batchEdit
import org.eclipse.tm4e.languageconfiguration.model.CommentRule

object CommentSystem {

  fun addSingleComment(commentRule: CommentRule?, text: Content) {
    text.insert(text.cursor.leftLine, 0, commentRule?.lineComment)
  }

  fun addBlockComment(commentRule: CommentRule?, text: Content) {
    val blockComment = commentRule?.blockComment

    val cursor = text.cursor
    if (cursor.isSelected) {
      text.batchEdit {
        // Insert multi-line comment at the beginning of the start line
        it.insert(cursor.leftLine, cursor.leftColumn, blockComment?.open)

        // Insert multi-line comment end at the end of the end line
        it.insert(cursor.rightLine, cursor.rightColumn, blockComment?.close)
      }
    }
  }
}
